rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,241 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::escape::escape_string;
+
+pub struct JsonObjectWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonObjectWriter<'a> {
+    pub fn new(output: &'a mut String) -> Self {
+        output.push('{');
+        Self {
+            json: output,
+            started: false,
+        }
+    }
+
+    /// Writes a null value with the given `key`.
+    pub fn null(&mut self, key: &str) -> &mut Self {
+        self.key(key);
+        self.json.push_str("null");
+        self
+    }
+
+    /// Writes the boolean `value` with the given `key`.
+    pub fn boolean(&mut self, key: &str, value: bool) -> &mut Self {
+        self.key(key);
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string `value` with the given `key`.
+    pub fn string(&mut self, key: &str, value: &str) -> &mut Self {
+        self.key(key);
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` (already converted to string) with the given `key`.
+    pub fn number(&mut self, key: &str, value: &str) -> &mut Self {
+        self.key(key);
+        self.json.push_str(value);
+        self
+    }
+
+    /// Starts an array with the given `key`.
+    pub fn start_array(&mut self, key: &str) -> JsonArrayWriter {
+        self.key(key);
+        JsonArrayWriter::new(&mut self.json)
+    }
+
+    /// Starts an object with the given `key`.
+    pub fn start_object(&mut self, key: &str) -> JsonObjectWriter {
+        self.key(key);
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the object.
+    pub fn finish(self) {
+        self.json.push('}');
+    }
+
+    fn key(&mut self, key: &str) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+
+        self.json.push('"');
+        self.json.push_str(&escape_string(key));
+        self.json.push_str("\":");
+    }
+}
+
+pub struct JsonArrayWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonArrayWriter<'a> {
+    pub fn new(output: &'a mut String) -> Self {
+        output.push('[');
+        Self {
+            json: output,
+            started: false,
+        }
+    }
+
+    /// Writes a null value to the array.
+    pub fn null(&mut self) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str("null");
+        self
+    }
+
+    /// Writes the boolean `value` to the array.
+    pub fn boolean(&mut self, value: bool) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string to the array.
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
```

Comment:
I _might_ consider a small optimization here for serializing base64 data (which I know can be a hot path).

In the case of base64, we know ahead of time that we don't need to perform any escaping so in that case we could do something like `string_unchecked` that skipped the escape phase.

Reply:
(here and other `string` methods)

Reply:
I like this idea.